File: 20230613_Tuesday/20230613.rb
# Leetcode: 2352. Equal Row and Column Pairs.
# https://leetcode.com/problems/equal-row-and-column-pairs/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 408 ms, faster than 17.39% of Ruby online submissions for Equal Row
# and Column Pairs.
# Memory Usage: 219.7 MB, less than 56.52% of Ruby online submissions for Equal
# Row and Column Pairs.
# 2023.06.13 Daily Challenge.
# @param {Integer[][]} grid
# @return {Integer}
def equal_pairs(grid)
  g = grid
  rows = g.map{|x| x.join(?,)}.tally # Was a bug here, a delimiter needed.
  cols = g.transpose.map{|x| x.join(?,)}.tally
  ans = 0 
  rows.keys.each do |hash|
    cols[hash] ||= 0
    ans += rows[hash] * cols[hash]
  end
  ans
end

File: 20230613_Tuesday/20230613.go
// Leetcode: 2352. Equal Row and Column Pairs.
// https://leetcode.com/problems/equal-row-and-column-pairs/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 218 ms, faster than 22.24% of Go online submissions for Equal Row
// and Column Pairs.
// Memory Usage: 8.7 MB, less than 22.91% of Go online submissions for Equal Row
// and Column Pairs.
// 2023.06.20 Updated.
package main

import (
	"strconv"
)

var r map[string]int // Rows.
var c map[string]int // Columns
var n int
var g [][]int

func equalPairs(grid [][]int) int {
	g = grid
	n = len(g)
	r = make(map[string]int)
	c = make(map[string]int)
	ans := 0
	fillRows()
	fillCols()
	for k, rv := range r {
		if cv, ok := c[k]; ok {
			ans += rv * cv
		}
	}
	return ans
}

func fillRows() {
	for i := 0; i < n; i += 1 {
		key := ""
		for j := 0; j < n; j += 1 {
			key += (strconv.Itoa(g[i][j]) + ",")
		}
		r[key] += 1
	}
}
func fillCols() {
	for j := 0; j < n; j += 1 {
		key := ""
		for i := 0; i < n; i += 1 {
			key += (strconv.Itoa(g[i][j]) + ",")
		}
		c[key] += 1
	}
}
